torch/testing/_internal/optests/generate_tests.py

```diff
@@ -0,0 +1,307 @@
+import functools
+
+import torch
+
+import torch.utils._pytree as pytree
+
+from torch._subclasses.schema_check_mode import SchemaCheckMode
+from torch.overrides import TorchFunctionMode
+from torch.testing._internal.optests import (
+    aot_autograd_check,
+    autograd_registration_check,
+    fake_check,
+)
+
+
+def safe_schema_check(op, args, kwargs):
+    args, kwargs = deepcopy_tensors(args, kwargs)
+    with SchemaCheckMode():
+        result = op(*args, **kwargs)
+        return result
+
+
+def safe_autograd_registration_check(op, args, kwargs):
+    # Don't perform autograd_registration_check if none of the inputs require grad.
+    if not pytree.tree_any_only(
+        torch.Tensor, lambda x: x.requires_grad, (args, kwargs)
+    ):
+        return
+    args, kwargs = deepcopy_tensors(args, kwargs)
+    return autograd_registration_check(op, args, kwargs)
+
+
+def safe_fake_check(op, args, kwargs):
+    args, kwargs = deepcopy_tensors(args, kwargs)
+    return fake_check(op, args, kwargs, dynamic_only=False)
+
+
+def safe_aot_autograd_check(op, args, kwargs, dynamic):
+    def func(*args, **kwargs):
+        args, kwargs = pytree.tree_map_only(torch.Tensor, torch.clone, (args, kwargs))
+        return op(*args, **kwargs)
+
+    # aot_autograd_check runs func(*args, **kwargs) multiple times
+    # and assumes `func` does not modify its inputs.
+    return aot_autograd_check(func, args, kwargs, dynamic, check_gradients="auto")
+
+
+# Test requirements
+# - The tests must have signature (op: OpOverload, args, kwargs)
+# - The tests must NOT mutate args, kwargs.
+# - The tests in this list must not be prefixes of each other. For example,
+#   having both "test_schema" and "test_schema_is_functional" is NOT OK.
+ALL_TESTS = {
+    "test_schema": safe_schema_check,
+    "test_autograd_registration": safe_autograd_registration_check,
+    "test_faketensor": safe_fake_check,
+    "test_aot_dispatch_static": functools.partial(
+        safe_aot_autograd_check,
+        dynamic=False,
+    ),
+    "test_aot_dispatch_dynamic": functools.partial(
+        safe_aot_autograd_check,
+        dynamic=True,
+    ),
+}
+
+
+def generate_opcheck_tests(
+    testcase, namespaces, failures_dict, additional_decorators, tests
+):
+    """Given an existing TestCase, use the existing tests to generate
```

Comment:
Is there any required structure on the TestCase you pass in here?

Reply:
It should be a subclass of unittest.TestCase. Other than that, no

Reply:
Just curious, will the Official(tm) custom ops docs talk about testing? / do you think this will/should eventually be public API?